ozaria/site/common/ozariaUtils.js

```diff
@@ -89,23 +100,47 @@ export const getLevelsDataByOriginals = (levels, levelOriginals) => {
 /**
  * Creates a level status map using the list of level session objects.
  * @param {Object[]} sessions - The list of level session objects.
- * @returns {Object} - Object with key as the level original id, and value as complete/started.
+ * @returns {Object} - Object with key as the level original id, and value = complete/started/capstone stage last 'completed'.
  */
 export const getLevelStatusMap = (sessions) => {
   const levelStatusMap = {}
   for (const session of sessions) {
     const levelOriginal = (session.level || session.get('level') || {}).original
-    if (levelOriginal && levelStatusMap[levelOriginal] !== 'complete') { // Don't overwrite a complete session with an incomplete one
-      if ((session.state || session.get('state') || {}).complete) {
+    const sessionState = session.state || session.get('state') || {}
+    const capstoneStage = sessionState.capstoneStage // current capstone stage (undefined if user is on stage 1)
+    const isLevelAreadyComplete = levelOriginal && (levelStatusMap[levelOriginal] === 'complete' || (capstoneStage && levelStatusMap[levelOriginal] === capstoneStage - 1))
+    if (!isLevelAreadyComplete) { // Don't overwrite a complete session with an incomplete one
+      if (capstoneStage > 1) {
```

Comment:
Is this always a number or can it be undefined?

Reply:
Yeah it is always a number (https://github.com/codecombat/codecombat/blob/master/app/schemas/models/level_session.coffee#L143).
It can be undefined when the user is on stage 1, and this condition should be false in that case.
To make it safer I will update it to `if (capstoneStage && capstoneStage > 1)`